### ISSUE TITLE
Principal Component Analysis (PCA)- Facial Recognition

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#12080]
Principal Component Analysis (PCA)- Facial Recognition 😀🤓😳🤠<br>
<br>
https://ift.tt/DbKhywN<br>
<br>
Hua Shi<br>
<br>

<div><div><div><div><div>
<div><h1>Principal Component Analysis (PCA)- Facial Recognition 😀🤓😳🤠</h1></div>
<figure></figure><p>Nowadays facial recognition is widely used and it is commonly used for ID verification — Face ID. For example, Apple’s Face ID is facial-recognition technology that launched on the iPhone X in 2017. Before your set up the Face ID, the iPhone will ‘scan’ your face from different angles and after you set it up, you are able to unlock your iPhones and it will also help you unlock some APPs. One more example, in some movies we saw that FIB, CIA, or similar investigation departments applied facial recognition technology to scan individuals’ pictures and obtain their information. So the question is how does facial recognition work? — The technology of facial recognition is based on the information of facial images. Facial identity features are extracted and compared with known faces to identify different faces.</p>
<h1>Definition</h1>
<p><a rel="noopener ugc nofollow" href="https://searchenterpriseai.techtarget.com/definition/facial-recognition" target="_blank">Facial recognition is a category of biometric software that maps an individual’s facial features mathematically and stores the data as a faceprint. The software uses deep learning algorithms to compare a live capture or digital image to the stored faceprint in order to verify an individual’s identity.</a></p>
<h1>Principal Component Analysis (PCA)</h1>
<p>PCA was invented in 1901 by <a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Karl_Pearson" target="_blank">Karl Pearson</a> who is the creator of <strong>Pearson’s</strong> Coefficient <strong>Correlation which is a </strong>mathematical method to measure the linear relationship between variables<strong>. </strong>After Karl Person, PCA has been developed many times. PCA is an algorithm of unsupervised learning and widely used to reduce the dimension of the data. <a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Principal_component_analysis" target="_blank">PCA is an orthogonal linear transformation which means that PCA can transform data into a new coordinate system such that the greatest variance by some scalar projection of the data comes to lie on the first principal component, the second greatest variance on the second </a><a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Principal_component_analysis" target="_blank">principal component</a><a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Principal_component_analysis" target="_blank">, and so on.</a> This is also called the process of dimensional reduction. The reason why it chooses the direction of bigger variance because it could significantly compress the dimension without losing a lot of information.</p>
<figure><figcaption><a rel="noopener ugc nofollow" href="https://www.neuraldesigner.com/blog/principal-components-analysis" target="_blank">By Pablo Martin, Artelnics.</a></figcaption></figure><h1>Code in Python</h1>
<h2>Data</h2>
<p>There are some <a rel="noopener ugc nofollow" href="https://analyticsindiamag.com/10-face-datasets-to-start-facial-recognition-projects/" target="_blank">datasets</a> for the facial recognition project. Here I just simply use the dataset from sklearn, and when you use this dataset please credit to AT&amp;T Laboratories Cambridge.</p>
<pre><span># Import necessary packages<br>import pandas as pd<br>import numpy as np<br>import matplotlib.pyplot as plt<br>import seaborn as sns<br>sns.set()<br>from sklearn.datasets import fetch_olivetti_faces</span><span># Load the data<br>data = fetch_olivetti_faces()<br>data.keys()<br>&gt;&gt;&gt; dict_keys(['data', 'images', 'target', 'DESCR'])</span><span># declare inputs, target and images<br>inputs=data.data<br>target=data.target<br>images=data.images</span><span>inputs.shape<br>&gt;&gt;&gt; (400, 4096)</span></pre>
<p>The data from sklearn is ready to use which means that we do not need to do the part of data cleansing and data engineering. After we load the data we can see the data is a dictionary and there are four different keys — The “target” is an integer from 0 to 39 indicating the identity of the person pictured. The “DESCR” contains the attribute information. The ‘data’ is the 2D array (400 *4096) transformed data from “image”. The “image” contains 400 different matrics and each matric’s shape is 64 by 64 which is the “digital expression” of image. So if we reshape the “image” we can get “data”. The picture below can explain more about “reshape”. Array 1 is just like our “image” data what I did here is I just reshaped the array 1 using each matric’s shape 3*3 then simply use <em>reshape</em> function to transform the matrics into a 2D array — array 2. The reason why I explain it because the data from sklean and it is already cleaned up, however, if we have raw image data, we need to use packages <em>os</em> and <em>cv2</em> to load, resize, and reshape the image data.</p>
<figure></figure><p>Let us take look at some images from this sklearn dataset.</p>
<pre><span>plt.figure(figsize=(20,25))<br>for i in range(50,70):<br> plt.subplot(4,5,i-49)<br> plt.imshow(data.images[i], cmap=plt.cm.gray)<br>plt.show()</span></pre>
<figure></figure><p>Each person represents one class and each class has 10 face images. All images for one person are different- some of them look at different directions and some of them wear glasses. However, after dimensional reduction, those images will also look different. Before we reduce the dimension, we also need to know how many components we need to set up.</p>
<h2>PCA &amp; Explained Variance</h2>
<pre><span># import library<br>from sklearn.decomposition import PCA<br>from sklearn.model_selection import train_test_split</span><span># train test split<br>x_train, x_test, y_train, y_test = train_test_split(inputs, target, random_state=365)<br># PCA transformation and reduce the dimension from 4096 to 50<br>pca = PCA(n_components=50, whiten=True)<br>X_train = pca.fit_transform(X_train)<br>X_test = pca.transform(X_test)</span><span>plt.plot(range(1,201), pca.explained_variance_ratio_.cumsum())<br>plt.title('Explained Variance',fontsize=15)<br>plt.xlabel('Number of Principle Components', fontsize=10)<br>plt.show()</span></pre>
<figure></figure><p><code><em>pca.explained_variance_ratio_</em></code>returns a vector of the variance explained by each dimension which also implies that the average distance between these points and the origin keeps growing when the dimension increases accordingly. Why this curve is important? Let us take a quick look at when <em>n_components </em>change.</p>
<pre><span># create a function to plot line graph to compare ratio - Explained Variance changes with different n component<br>def EV(n):<br> pca = PCA(n_components=n, whiten=True)<br> X_train = pca.fit_transform(x_train)<br> X_test = pca.transform(x_test)<br> #plot<br> plt.plot(range(1,n+1), pca.explained_variance_ratio_.cumsum())<br> plt.title('Explained Variance',fontsize=15)<br> plt.xlabel('Number of Principle Components : {}'.format(n), fontsize=10)</span><span># use the function above and try use n_component=100,200,and 300<br>plt.figure(figsize=(15,5))<br>plt.subplot(1,3,1)<br>EV(100)<br>plt.subplot(1,3,2)<br>EV(200)<br>plt.subplot(1,3,3)<br>EV(300)</span></pre>
<figure></figure><p>We can see that the ratio increased a lot when changes of <em>n_components are </em>from 50 to 300 dimensions, which means when data is compressed into a certain number of dimensions, those dimensions can capture 99% of the overall variance. Briefly speaking, our inputs data has 4096 dimensions and if we compress the inputs data into 200 dimensions, we still can obtain the most information from inputs data. Using this visualization method can help us to choose the number of <em>components </em>better.</p>
<h1>Modeling</h1>
<p>Support Vector Machines (SVM) are one of the most useful techniques in classification problems. SVM algorithm is commonly used for face recognition analysis. The code below is a simple SVM model. If you want to obtain higher accuracy you can apply GridSearch to tune the parameters. Even though the best parameters were not applied in this model, the test accuracy already reached around 96%.</p>
<pre><span># Credit to Flatiron School PCA lecture <br># fit the model<br>clf = svm.SVC(C=10, gamma=0.001)<br>%timeit clf.fit(X_train, y_train)</span><span># obtian the accuracy<br>accuracy_train= clf.score(X_train, y_train)<br>accuracy_test= clf.score(X_test, y_test)<br>print('Accuracy - train data: {}'.format(accuracy_train))<br>print('Accuracy - test data : {}'.format( accuracy_test))</span><span># prediction<br>test_pre=clf.predict(X_test)<br>train_pre=clf.predict(X_train)</span><span># F1 _score<br>from sklearn.metrics import f1_score<br>f1_train=f1_score(y_train,train_pre,average='weighted')<br>f1_test=f1_score(y_test,test_pre,average='weighted')<br>print("f1 score - test data : {}" .format(f1))<br>&gt;&gt;&gt; Accuracy - train data: 1.0<br>&gt;&gt;&gt; Accuracy - test data : 0.96</span></pre>
<h2>Mini-Backed Project for Face Recognition</h2>
<pre><span># create a simple mini-project - recognize face image</span><span>def Actual_VS_Prediction():<br> n=int(input("Please chose a face image#: "))# first you can choose any image that you want "scan"<br> test_compressed = pca.transform(x_test)<br> plt.subplot(1,2,1)<br> plt.imshow(pca.inverse_transform(test_compressed[y_test[int(n)]]).reshape(64,64), cmap=plt.cm.gray)<br> plt.title("Actual Image")<br> # prediction<br> plt.subplot(1,2,2)<br> plt.imshow(pca.inverse_transform(test_compressed[test_pre[int( n)]]).reshape(64,64), cmap=plt.cm.gray)<br> plt.title("Predicted Image")</span><span>if test_pre[n] == y_test[n]:<br> print(" Yay! The image matches! ")<br> else: <br> print(" Sorry! The image does not match!")<br> plt.show()</span></pre>
<figure></figure><h2>How to load your own image data and reshape them?</h2>
<p>The first thing we need to do is to import packages — os and cv2. Then we need to find the path of the image folder. Next, we can choose to take a look at the first image from the folder before PCA.</p>
<pre><span># import packages<br>import cv2<br>import os</span><span># find the path of your image folder<br>path='copy and paste the path of your image folder'</span><span># check one image from the folder<br>for img in os.listdir(path):<br> img_array=cv2.imread(os.path.join(path,img),cv2.IMREAD_GRAYSCALE)<br> plt.imshow(img_array,cmap="gray")<br> plt.show()<br> break</span></pre>
<figure><figcaption>This is my selfie 😊</figcaption></figure><pre><span># create a function to create,load,resize, and re-color the data<br>def create_data(path):<br> list_new_array=[]<br> for img in os.listdir(path):<br> try:<br> img_array=cv2.imread(os.path.join(path,img),cv2.IMREAD_GRAYSCALE)<br> size=100<br> new_array=cv2.resize(img_array,(size,size))<br> list_new_array.append(new_array)<br> except Exception as e:<br> pass<br> return list_new_array</span></pre>
<p>Now it is time to reshape the data!! I already explained the idea behind the code in the previous part in this blog. Let us take a look at the code. “hua” and “hua_reshape” are just like the sklearn face image data “image” and “data”. Then you can apply PCA and SVM with this data.</p>
<pre><span># I called the data is "hua" (my name)<br>hua=np.array(create_data(path))<br>hua_reshape=hua.reshape(8, 100*100).astype('float32')</span><span>plt.figure(figsize=(18,12))<br>for i in [3,4,5,6,7,8]:<br> plt.subplot(2,3,i-2)<br> n_feats = i<br> pca = PCA(n_components=n_feats)<br> pca.fit(hua_reshape)<br> compressed = pca.transform(hua_reshape)<br> plt.title('Recovered Image from\n{} principle components'.format(n_feats))<br> plt.imshow(pca.inverse_transform(compressed[6]).reshape(100,100), cmap=plt.cm.gray)</span></pre>
<figure><figcaption>Recovered Image from different principle components</figcaption></figure>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium <br>
<br>
March 24, 2025 at 01:53PM
